crates/sui-adapter/src/temporary_store.rs

```diff
@@ -83,42 +70,55 @@ impl<S> TemporaryStore<S> {
 
     // Helpers to access private fields
     pub fn objects(&self) -> &BTreeMap<ObjectID, Object> {
-        &self.objects
-    }
-
-    pub fn written(&self) -> &BTreeMap<ObjectID, Object> {
-        &self.written
+        &self.input_objects
     }
 
     pub fn deleted(&self) -> &BTreeMap<ObjectID, (SequenceNumber, DeleteKind)> {
         &self.deleted
     }
 
     /// Break up the structure and return its internal stores (objects, active_inputs, written, deleted)
-    pub fn into_inner(self) -> InnerTemporaryStore {
+    pub fn into_inner(self) -> (InnerTemporaryStore, Vec<Event>) {
         #[cfg(debug_assertions)]
         {
             self.check_invariants();
         }
-        into_inner!(self)
+        let written = self
+            ._written
+            .into_iter()
+            .map(|(id, obj)| (id, (obj.compute_object_reference(), obj)))
+            .collect();
+        (
+            InnerTemporaryStore {
+                objects: self.input_objects,
+                mutable_inputs: self.mutable_input_refs,
+                written,
+                deleted: self.deleted,
+            },
+            self.events,
+        )
     }
 
     /// For every object from active_inputs (i.e. all mutable objects), if they are not
     /// mutated during the transaction execution, force mutating them by incrementing the
     /// sequence number. This is required to achieve safety.
     /// We skip the gas object, because gas object will be updated separately.
     pub fn ensure_active_inputs_mutated(&mut self, gas_object_id: &ObjectID) {
-        for (id, _seq, _) in &self.mutable_inputs {
+        let mut to_be_updated = vec![];
+        for (id, _seq, _) in &self.mutable_input_refs {
             if id == gas_object_id {
                 continue;
             }
-            if !self.written.contains_key(id) && !self.deleted.contains_key(id) {
-                let mut object = self.objects[id].clone();
+            if !self._written.contains_key(id) && !self.deleted.contains_key(id) {
+                let mut object = self.input_objects[id].clone();
                 // Active input object must be Move object.
                 object.data.try_as_move_mut().unwrap().increment_version();
-                self.written.insert(*id, object);
+                to_be_updated.push(object);
             }
         }
+        for object in to_be_updated {
+            self.write_object(object);
```

Comment:
Cannot easily tell why we cannot inline this call above when we update the vector? Maybe comment?